src/Infrastructure/TimeMgr/src/ESMCI_Alarm.C

```diff
@@ -129,59 +128,61 @@ int Alarm::count=0;
       sprintf(alarm->name, "Alarm%3.3d", alarm->id);
     }
 
-    if (ringTime != ESMC_NULL_POINTER) {
-      alarm->ringTime = alarm->prevRingTime = alarm->firstRingTime = *ringTime;
-    }
+    if (ringTime != ESMC_NULL_POINTER) 
+      alarm->ringTime = *ringTime;
+    else
+      alarm->ringTime = clock->currTime;
+
     if (ringInterval != ESMC_NULL_POINTER) {
-      alarm->ringInterval = *ringInterval;
+      TimeInterval zeroTimeInterval(0,0,1,0,0,0);
+      if(*ringInterval < zeroTimeInterval){
+        ESMC_LogDefault.MsgFoundError(ESMC_RC_ARG_BAD,
+          "; Cannot set negative ringInterval", ESMC_CONTEXT, rc);
+        return 0;
+      }
 
-      // if ringTime not specified, calculate
-      //   ringTime from the current clock time
+      if(*ringInterval < clock->timeStep) {
+        ESMC_LogDefault.MsgFoundError(ESMC_RC_ARG_BAD,
+          "; alarm Interval cannot be less than clock interval", ESMC_CONTEXT, rc);
+        return 0;
+      }
 
-      // TODO: handle case where *ringTime < clock->currTime;
-      //       same or similar to refTime
-      
-      if (ringTime == ESMC_NULL_POINTER) {
-        // works for positive or negative ringInterval
-        alarm->ringTime = clock->currTime + alarm->ringInterval;
-        alarm->prevRingTime = alarm->firstRingTime = alarm->ringTime;
+      if( ( *ringInterval == clock->timeStep) ){
+        int n = (alarm->ringTime - clock->currTime)/(*ringInterval);
+        //std::cout << n << std::endl;
+        //ringTime->print(); 
+        //clock->currTime.print(); 
+        //ringInterval->print();
+        if(alarm->ringTime != (clock->currTime + n*(*ringInterval)) ){
+          ESMC_LogDefault.MsgFoundError(ESMC_RC_ARG_BAD,
+            "; alarm Interval is the same as clock interval, ringTime cannot be misaligned with clock time, alarm will never ring", ESMC_CONTEXT, rc);
+          return 0;
+        }
       }
+
+      // We do not check LCM because any two rational numbers are guranteed to have
+      // LCM albeit it may be very large.
+      alarm->ringInterval = *ringInterval;
+
+    }else{
+      // A negative ringInterval indicates a one shot alarm at ringTime only
+      TimeInterval zeroTimeInterval(0,0,1,0,0,0);
+      if(clock->timeStep < zeroTimeInterval)
+        alarm->ringInterval = clock->timeStep;
+      alarm->ringInterval = -clock->timeStep;
     }
```

Comment:
- [ ] This seems problematic, particularly given the code in `ringerOff` that decrements `ringTime` by `ringInterval`. (See also https://github.com/esmf-org/esmf/pull/340#issuecomment-2566736806.) I think it would be better and safer to have a boolean flag that says whether this alarm rings on an interval or is a one-shot alarm.